internal/datum/credentials.go

```diff
@@ -0,0 +1,95 @@
+package datum
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+	"path"
+
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/jwt"
+)
+
+func DefaultTokenSource(ctx context.Context) (oauth2.TokenSource, error) {
+	var creds *credentialsFile
+	var err error
+
+	// Check if the user specified the location through the well-known environment
+	// variable.
+	if filename := os.Getenv("DATUM_CREDENTIALS"); filename != "" {
+		creds, err = readCredentialsFile(ctx, filename)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read credentials file: %w", err)
+		}
+	}
+
+	// Check if it's in the well-known file location.
+	homeDir, err := os.UserHomeDir()
+	if err != nil {
+		return nil, err
+	}
+
+	wellKnownFile := path.Join(homeDir, ".config", "datum", "application_credentials.json")
+	if _, err := os.Stat(wellKnownFile); err == nil {
+		creds, err = readCredentialsFile(ctx, wellKnownFile)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read application credentials from home directory: %w", err)
+		}
+	}
+
+	if creds == nil {
+		return nil, fmt.Errorf("could not find default application credentials")
+	}
+
+	audience := creds.Audience
+	if audience == "" {
+		audience = "https://auth.datum.net"
+	}
+
+	ctx = context.WithValue(ctx, oauth2.HTTPClient, &http.Client{
+		Transport: &wrappedRoundTripper{
+			Transport: http.DefaultTransport,
+			AdditionalFields: map[string]string{
+				"client_id": "jwt-client",
+				"audience":  audience,
+			},
+		},
+	})
```

Comment:
Can you help me understand why these are necessary? I see that `Audience` is defined in the `jwt.Config` below.

Reply:
The `client_id` is required to let Ory Hydra know which client is attempting to authenticate with the token endpoint. There's a public OAuth2 client setup with the client ID `jwt-client` that supports JWT Bearer token exchange authentication which allows the CLI to sign its own JWT token and exchange it for an access token. I'd prefer to move this out of the client and into the server. Maybe I should slap a TODO on there to make that clear? 

The audience value is another scenario where the server side can't inject audience values into the resulting access token that's provisioned. The Audience used for the JWT we sign before sending to the token endpoint MUST be the token endpoint otherwise the request is rejected. We could technically remove the audience here and have any OIDC provider trust the token endpoint as the audience value, but having a generic audience feels more correct.

Reply:
> The client_id is required to let Ory Hydra know which client is attempting to authenticate with the token endpoint. There's a public OAuth2 client setup with the client ID jwt-client that supports JWT Bearer token exchange authentication which allows the CLI to sign its own JWT token and exchange it for an access token. I'd prefer to move this out of the client and into the server. Maybe I should slap a TODO on there to make that clear?

I see that [RFC 7523](https://datatracker.ietf.org/doc/rfc7523/)  states that authentication of the client is optional, I'm guessing that Hydra requires it?

>  Authentication of the client is optional, as described in
   Section 3.2.1 of OAuth 2.0 [RFC6749] and consequently, the
   "client_id" is only needed when a form of client authentication that
   relies on the parameter is used.

Looks like a bug report was filed with Hydra about this topic a few days ago (assuming I'm interpreting how we're using this correctly): https://github.com/ory/hydra/issues/3903

> The audience value is another scenario where the server side can't inject audience values into the resulting access token that's provisioned. The Audience used for the JWT we sign before sending to the token endpoint MUST be the token endpoint otherwise the request is rejected. We could technically remove the audience here and have any OIDC provider trust the token endpoint as the audience value, but having a generic audience feels more correct.

Does the `Audience` field in the JWT config not accomplish this? I see that it results in an `aud` claim being added to the token that's sent.

Reply:
> I see that [RFC 7523](https://datatracker.ietf.org/doc/rfc7523/) states that authentication of the client is optional, I'm guessing that Hydra requires it?

Correct, this is just a requirement that Ory Hydra has in place.

> Looks like a bug report was filed with Hydra about this topic a few days ago (assuming I'm interpreting how we're using this correctly). 

That looks to be the same issue. I came across another GitHub issue that highlighted this problem when I was investigating this several weeks ago but I'm having trouble finding it again.

> Does the Audience field in the JWT config not accomplish this? I see that it results in an aud claim being added to the token that's sent.

Ory Hydra will only accept an audience value that matches the token URL being used and the access token that's issued will automatically include that as the audience. As long as we're okay with setting the trusted audience to the token URL (e.g. `https://auth.datum.net/oauth2/token`) in any system relying on Ory Hydra as a OIDC provider, we can remove the custom audience that's being supplied here.

Reply:
> As long as we're okay with setting the trusted audience to the token URL (e.g. https://auth.datum.net/oauth2/token) in any system relying on Ory Hydra as a OIDC provider, we can remove the custom audience that's being supplied here.

I don't think we'd want a static audience, but audiences for different system components, like a specific deployment of the datum-apiserver. I think what you have is good for now, we'll iterate.

Reply:
> I don't think we'd want a static audience, but audiences for different system components, like a specific deployment of the datum-apiserver.

Agreed there.